src/mpid/ch4/shm/posix/posix_am.h

```diff
@@ -411,4 +333,116 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_POSIX_am_send_hdr_reply(MPIR_Context_id_t con
+        }
+
+        MPIDIG_am_send_async_init(sreq, datatype, data_sz);
+        need_packing = dt_contig ? false : true;
+    } else {
+        /* if we are issuing as deferred operation, use the msg_hdr in sreq */
+        msg_hdr_p = MPIDI_POSIX_AMREQUEST_HDR(sreq, msg_hdr);
+        need_packing = MPIDI_POSIX_AMREQUEST_HDR(sreq, is_contig) ? false : true;
+    }
+
+    if (!issue_deferred && MPIDI_POSIX_global.postponed_queue) {
+        goto fn_deferred;
+    }
+
+    rc = MPIDI_POSIX_eager_send(grank, data, count, datatype, am_hdr, am_hdr_sz, need_packing,
+                                &msg_hdr_p, MPIDIG_am_send_async_get_data_sz_left(sreq),
+                                MPIDIG_am_send_async_get_offset(sreq), &send_size);
+
+    if (rc == MPIDI_POSIX_NOK) {
+        if (!issue_deferred) {
+            goto fn_deferred;
+        } else {
+            goto fn_exit;
+        }
+    }
+
+    MPL_DBG_MSG_FMT(MPIDI_CH4_DBG_GENERAL, VERBOSE,
+                    (MPL_DBG_FDEST,
+                     "issue seg for req handle=0x%x send_size %ld", sreq->handle, send_size));
+    MPIDIG_am_send_async_issue_seg(sreq, send_size);
+    MPIDIG_am_send_async_finish_seg(sreq);
+
+    /* if there IS MORE DATA to be sent and we ARE NOT called for issue deferred op, enqueue.
+     * if there NO MORE DATA and we ARE called for issuing deferred op, pipeline is done, dequeue
+     * skip for all other cases */
+    if (!MPIDIG_am_send_async_is_done(sreq)) {
+        if (!issue_deferred) {
+            goto fn_deferred;
+        }
+    } else {
+        /* all segments are sent, complete regularly and dequeue (if was issuing deferred op) */
+        if (issue_deferred) {
+            MPIDI_POSIX_am_request_header_t *curr_req_hdr = MPIDI_POSIX_global.postponed_queue;
+            DL_DELETE(MPIDI_POSIX_global.postponed_queue, curr_req_hdr);
+
+            MPL_free(MPIDI_POSIX_AMREQUEST_HDR(sreq, pack_buffer));
+            MPIDI_POSIX_AMREQUEST_HDR(sreq, pack_buffer) = NULL;
+        }
+        mpi_errno = MPIDIG_global.origin_cbs[handler_id] (sreq);
+        MPIR_ERR_CHECK(mpi_errno);
+    }
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_POSIX_DO_AM_ISEND);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+  fn_deferred:
+    MPL_DBG_MSG_FMT(MPIDI_CH4_DBG_GENERAL, VERBOSE,
+                    (MPL_DBG_FDEST, "deferred posix_am_isend req handle=0x%x", sreq->handle));
+
+    mpi_errno = MPIDI_POSIX_am_enqueue_request(am_hdr, am_hdr_sz, handler_id, grank, msg_hdr,
+                                               msg_hdr_p, NULL, 0, data_sz, sreq);
+    MPIDI_POSIX_AMREQUEST_HDR(sreq, comm) = comm;
+    MPIDI_POSIX_AMREQUEST_HDR(sreq, rank) = rank;
+    MPIDI_POSIX_AMREQUEST_HDR(sreq, buf) = data;
+    MPIDI_POSIX_AMREQUEST_HDR(sreq, datatype) = datatype;
+    MPIDI_POSIX_AMREQUEST_HDR(sreq, count) = count;
+    MPIDI_POSIX_AMREQUEST_HDR(sreq, kind) = kind;
+    MPIDI_POSIX_AMREQUEST_HDR(sreq, is_contig) = dt_contig;
```

Comment:
Is this part separate from the eager module interface refactor? If so, we should split the commits.

Reply:
It is not. It is a necessary change related to the interface refactoring.